### PR TITLE
Grammar caching and validation for RNG-based topics

### PR DESCRIPTION
**Pull request by raducoravu**
FIxes for #3661 and #3253
Signed-off-by: Radu Coravu <radu_coravu@sync.ro>

## Description
Added grammar caching for Relax NG based DITA topics and maps. Makes publishing faster for such RNG-based DITA projects.
Added validation for Relax NG based DITA topics and maps. Reports in the console validation errors if the topics are invalid according to the RNG schema.

## Motivation and Context
FIxes for #3661 and #3253

## How Has This Been Tested?
Tested manually with a small RNG-based DITA project on Mac OS.

## Type of Changes
Speed performance improvement and removed limitation which did not present validation errors for RNG-based topics.
- New feature _(non-breaking change which adds functionality)_

## Documentation and Compatibility
Maybe mention in the release notes:
Added grammar caching for Relax NG based DITA topics and maps. Makes publishing faster for such RNG-based DITA projects.
Added validation for Relax NG based DITA topics and maps. Reports in the console validation errors if the topics are invalid according to the RNG schema.

- What documentation changes are needed for this feature?
Not sure if the limitation that RNG-based publishing did not validate was mentioned anywhere in the user's guide.
- Will this change affect backwards compatibility or other users' overrides?
No

## Checklist
- My code follows the code style of this project.
    -  <https://github.com/dita-ot/dita-ot/wiki/Java-Coding-Conventions>
    -  <https://github.com/dita-ot/dita-ot/wiki/XSLT-Coding-Conventions>
    -  <https://github.com/dita-ot/docs/wiki/coding-guidelines>
- I have updated the unit tests to reflect the changes in my code.

### CODE REVIEW

File: src/main/java/org/ditang/relaxng/defaults/RelaxDefaultsParserConfiguration.java [hunk 1]
```diff
@@ -149,5 +158,23 @@ protected void insertRelaxDefaultsComponent() {
...
+  public void setFeature(String featureId, boolean state) throws XMLConfigurationException {
+    if(XERCES_SCHEMA_VALIDATION.equals(featureId)) {
+      relaxNGValidation = state;
+    } else {
+      super.setFeature(featureId, state);
+    }
+  }
   
+  @Override
+	public void setProperty(String propertyId, Object value) throws XMLConfigurationException {
```
Comment (jelovirt) [inline src/main/java/org/ditang/relaxng/defaults/RelaxDefaultsParserConfiguration.java:174]: The indentation in this PR uses both 2 and 4 spaces. Could you use 4 spaces for all indents?

File: src/main/java/org/ditang/relaxng/defaults/RelaxNGDefaultValues.java [hunk 2]
```diff
@@ -113,10 +138,15 @@ List<Attribute> getDefaultAttributes(String elementLocalName,
      *      java.lang.String, java.lang.String, java.lang.String,
      *      java.lang.String)
      */
+    @Override
     public void defaultValue(String elementLocalName, String elementNamespace,
         String attributeLocalName, String attributeNamepsace, String value) {
       String key = getKey(elementLocalName, elementNamespace);
-      List<Attribute> list = defaults.computeIfAbsent(key, k -> new ArrayList<>());
+      List<Attribute> list = defaults.get(key);
```
Comment (jelovirt) [inline src/main/java/org/ditang/relaxng/defaults/RelaxNGDefaultValues.java:145]: Doesn't this change just change the code from `computeIfAbsent` to doing the exact same thing "manually"?

File: src/main/java/org/ditang/relaxng/defaults/pool/RNGDefaultsEnabledGrammarPool.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
+/*
+ * This file is part of the DITA Open Toolkit project.
+ *
+ * Copyright 2011 George Bina
```
Comment (jelovirt) [inline src/main/java/org/ditang/relaxng/defaults/pool/RNGDefaultsEnabledGrammarPool.java:4]: Shouldn't this be 
```suggestion
 * Copyright 2022 Radu Coravu
```

File: src/main/java/org/ditang/relaxng/defaults/pool/RNGDefaultsEnabledSynchronizedXMLGrammarPoolImpl.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2019 Syncro Soft SRL - All Rights Reserved.
+ *
+ * This file contains proprietary and confidential source code.
+ * Unauthorized copying of this file, via any medium, is strictly prohibited.
+ */
```
Comment (jelovirt) [inline src/main/java/org/ditang/relaxng/defaults/pool/RNGDefaultsEnabledSynchronizedXMLGrammarPoolImpl.java:6]: Wrong copyright or if this needs to be retained, then this PR cannot be merged.

File: src/main/java/org/ditang/relaxng/defaults/pool/RNGDefaultsEnabledSynchronizedXMLGrammarPoolImpl.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+public class RNGDefaultsEnabledSynchronizedXMLGrammarPoolImpl
+    extends org.dita.dost.util.XMLGrammarPoolImplUtils implements RNGDefaultsEnabledGrammarPool {
+  
+  /**
+   * Caches Relax NG default values based on the URL pointing to the RNG schemas.
+   */
+  private final Map<String, RelaxNGDefaultValues> rngDefaultValues = new HashMap<>();
+  
+  /**
+   * @see ro.sync.xml.parser.RNGDefaultsEnabledGrammarPool#getRngDefaultValues(java.lang.String)
```
Comment (jelovirt) [inline src/main/java/org/ditang/relaxng/defaults/pool/RNGDefaultsEnabledSynchronizedXMLGrammarPoolImpl.java:28]:
```suggestion
   * @see org.ditang.relaxng.defaults.pool.RNGDefaultsEnabledGrammarPool#getRngDefaultValues(java.lang.String)
```

File: src/main/java/org/ditang/relaxng/defaults/pool/RNGDefaultsEnabledSynchronizedXMLGrammarPoolImpl.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+   */
+  @Override
+  public synchronized RelaxNGDefaultValues getRngDefaultValues(String systemID) {
+    return rngDefaultValues.get(systemID);
+  }
+  
+  /**
+   * @see ro.sync.xml.parser.RNGDefaultsEnabledGrammarPool#putRngDefaultValues(java.lang.String, org.ditang.relaxng.defaults.RelaxNGDefaultValues)
```
Comment (jelovirt) [inline src/main/java/org/ditang/relaxng/defaults/pool/RNGDefaultsEnabledSynchronizedXMLGrammarPoolImpl.java:36]:
```suggestion
   * @see org.ditang.relaxng.defaults.pool.RNGDefaultsEnabledGrammarPool#putRngDefaultValues(java.lang.String, org.ditang.relaxng.defaults.RelaxNGDefaultValues)
```

**Review by jelovirt — CHANGES_REQUESTED** [review]
Are there any unit tests that test that this code is run?

File: src/main/java/org/ditang/relaxng/defaults/RelaxNGDefaultValues.java [hunk 7]
```diff
@@ -113,10 +138,15 @@ List<Attribute> getDefaultAttributes(String elementLocalName,
...
+      if (list == null) {
+        list = new ArrayList<Attribute>();
+        defaults.put(key, list);
+      }
```
Comment (jelovirt) [inline src/main/java/org/ditang/relaxng/defaults/RelaxNGDefaultValues.java:149]: Doesn't this change just change the code from `computeIfAbsent` to doing the exact same thing "manually"?

**Review by jelovirt — CHANGES_REQUESTED** [review]

**Comment by chrispy-snps** [comment]
@raducoravu - thank you for contributing this!

The runtime improvement is significant:

```
$ time ~/git/dita-ot/src/main/bin/dita -i dita/ptug.ditamap -f html5 -o out1
real    4m30.200s
user    3m46.234s
sys     1m41.484s
$ time ~/git/dita-ot-cached/src/main/bin/dita -i dita/ptug.ditamap -f html5 -o out2
real    0m59.383s
user    1m18.000s
sys     0m32.313s
```

The baseline and RelaxNG-cached output directories are identical.

I confirmed that both DTD and RelaxNG topics now show similar-but-different validation messages:

```
 [gen-list] [DOTJ013E][ERROR] Failed to parse the referenced file 'file:.../topics/introduction-dtd.dita'.: file:.../topics/introduction-dtd.dita Line 10:The content of element type "topic" must match "(title,titlealts?,(shortdesc|abstract)?,prolog?,body?,related-links?,topic*)".

 [gen-list] [DOTJ013E][ERROR] Failed to parse the referenced file 'file:.../topics/introduction-rng.dita'.: file:.../topics/introduction-rng.dita Line 4:element "prolog" not allowed yet; missing required element "title"
```

The transformation behavior was consistent between DTD and RelaxNG in that the invalid topic was excluded from both final transformation results.

**Comment by raducoravu** [comment]
@chrispy-snps thanks for the tests, out of curiosity can you also post the times you get using the DITA OT bundled with Oxygen 24.1? It should have the fix included but I discovered that the fix may not have fully worked with it.

**Comment by raducoravu** [comment]
I added 2 auto tests to check that the grammar pool is used (but contains only unique schemas):
org.dita.dost.IntegrationTest.testRngGrammarPool()
org.dita.dost.IntegrationTest.testRngNoGrammarPool()

And 2 auto tests that the validate parameter is used:
org.dita.dost.IntegrationTestPreprocess1.testRngGrammarPoolValidate()
org.dita.dost.IntegrationTestPreprocess1.testRngGrammarPoolNoValidate()

I cannot add this latter tests to the common IntegrationTest because of this existing issue with preprocess2:
https://github.com/dita-ot/dita-ot/issues/3281
"preprocess2" does not break the parsing of an XML resource if the error handler returns exceptions for it. "preprocess" works correctly.

**Comment by chrispy-snps** [comment]
@raducoravu - The previous testcase did not show much difference, 59s for the cached DITA-OT versus 62s for the Oxygen Publishing Engine 2022041502.

I tried a different testcase with more (2300) smaller files. The results are as follows, averaged across several runs and with validation disabled for consistency:

* 262s for DITA-OT, cached
* 310s for Oxygen Publishing Engine 2022041502
* 971s for DITA-OT 3.7.2

**Review by jelovirt — APPROVED** [review]

**Comment by raducoravu** [comment]
Thanks @jelovirt

**Pull request merged**